### PR TITLE
support pushing coalesce and casewhen down to kv

### PR DESCRIPTION
**Pull request by XuHuaiyu**
This commit supports pushing coalesce and casewhen down to kv.
PTAL @hanfei1991 @coocood @shenli @zimulala @tiancaiamao

### CODE REVIEW

File: plan/expr_to_pb.go [hunk 1]
```diff
@@ -127,6 +127,8 @@ func scalarFuncToPBExpr(client kv.Client, expr *expression.ScalarFunction) *tipb
 		return arithmeticalFuncToPBExpr(client, expr)
 	case ast.AndAnd, ast.OrOr, ast.UnaryNot:
 		return logicalFuncToPBExpr(client, expr)
+	case ast.Case, "coalesce":
```
Comment (hanfei1991) [inline plan/expr_to_pb.go]: how about change "coalesce" to ast.Coalesce.

File: distsql/xeval/eval_other_funcs_test.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	"github.com/pingcap/tidb/util/types"
+	"github.com/pingcap/tipb/go-tipb"
+)
+
+func (s *testEvalSuite) TestEvalCaseWhen(c *C) {
+	colID := int64(1)
+	row := make(map[int64]types.Datum)
+	row[colID] = types.NewIntDatum(100)
+	xevaluator := &Evaluator{Row: row}
+	cases := []struct {
```
Comment (hanfei1991) [inline distsql/xeval/eval_other_funcs_test.go]: add tests for NULL

File: distsql/xeval/eval_other_funcs.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xeval
+
+import (
+	"github.com/pingcap/tidb/_vendor/src/github.com/juju/errors"
```
Comment (hanfei1991) [inline distsql/xeval/eval_other_funcs.go]: don't import vendor

File: plan/expression_rewriter.go [hunk 4]
```diff
@@ -666,7 +666,7 @@ func (er *expressionRewriter) caseToExpression(v *ast.CaseExpr) {
 		if v.ElseClause != nil {
 			args = append(args, er.ctxStack[stkLen-1])
 		}
-		argsLen++
+		argsLen++ // for trimming the value element later.
```
Comment (zimulala) [inline plan/expression_rewriter.go]: Remove `.` or complete the comment.
The same as line358 in plan_test.go

**Comment by XuHuaiyu** [comment]
@zimulala @coocood @hanfei1991 PTAL

**Comment by XuHuaiyu** [comment]
PTAL @hanfei1991 @coocood @zimulala

File: plan/expr_to_pb.go (hunk 1 above)
Comment (hanfei1991) [inline plan/expr_to_pb.go]: change "coalesce" to ast.Coalesce.

Comment (hanfei1991) [inline plan/expr_to_pb.go]: change "coalesce" to ast.Coalesce.

File: plan/expr_to_pb.go [hunk 5]
```diff
@@ -369,3 +371,25 @@ func aggFuncToPBExpr(client kv.Client, aggFunc expression.AggregationFunction) *
 	}
 	return &tipb.Expr{Tp: tp, Children: children}
 }
+
+func builtinFuncToPBExpr(client kv.Client, expr *expression.ScalarFunction) *tipb.Expr {
+	var tp tipb.ExprType
+	switch expr.FuncName.L {
+	case ast.Case:
+		tp = tipb.ExprType_Case
+	case "coalesce":
```
Comment (hanfei1991) [inline plan/expr_to_pb.go]: ditto.

File: plan/plan_test.go [hunk 6]
```diff
@@ -348,6 +350,80 @@ func (s *testPlanSuite) TestLogicOpsPushDown(c *C) {
 	}
 }
 
+// TestBuiltinFuncsPushDown tests whether logic operators been pushed down successfully.
+func (s *testPlanSuite) TestBuiltinFuncsPushDown(c *C) {
+	defer testleak.AfterTest(c)()
+	cases := []struct {
+		sql    string
+		cond   string // readable expressions
+		exprPB string // Marshall result of conditions been pushed down.
```
Comment (hanfei1991) [inline plan/plan_test.go]: that be pushed down.

**Comment by XuHuaiyu** [comment]
@hanfei1991 PTAL

**Comment by hanfei1991** [comment]
LGTM

File: distsql/xeval/eval_other_funcs.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+		if err != nil {
+			return d, errors.Trace(err)
+		}
+		return ans, nil
+	}
+	return d, nil
+}
+
+func (e *Evaluator) evalCoalesce(expr *tipb.Expr) (d types.Datum, err error) {
+	for _, child := range expr.Children {
```
Comment (shenli) [inline distsql/xeval/eval_other_funcs.go]: for _, child := range expr.Children {
 		d, err = e.Eval(child)
 		if err != nil {
 			break
 		}
 		if x.IsNull() {
 			break
 		}
 	}
 +	return d, errors.Trace(err)
}

File: distsql/xeval/eval_test.go [hunk 8]
```diff
@@ -276,25 +276,25 @@ func (s *testEvalSuite) TestEval(c *C) {
...
+		case types.Datum:
+			expr.Children[i] = datumExpr(x)
+		case *tipb.Expr:
+			expr.Children[i] = x
+		}
 	}
 	return expr
 }
 
+func binaryExpr(left, right interface{}, tp tipb.ExprType) *tipb.Expr {
```
Comment (shenli) [inline distsql/xeval/eval_test.go]: Why we need this function?

**Comment by XuHuaiyu** [comment]
@shenli PTAL

**Comment by shenli** [comment]
LGTM
